server/app/services/applicant/ApplicantServiceImpl.java

```diff
@@ -358,6 +367,139 @@ public ImmutableList<Application> getAllApplications() {
+                .collect(Collectors.toMap(ProgramDefinition::adminName, pdef -> pdef)));
+
+    // When new versions of Programs are created, they have distinct IDs but retain the
+    // same adminName. In order to find the most recent draft / active application,
+    // we first group by the unique name rather than the ID.
+    Map<String, Map<LifecycleStage, Optional<Application>>> mostRecentApplicationsByProgram =
+        applications.stream()
+            .collect(
+                Collectors.groupingBy(
+                    a -> {
+                      return a.getProgram().getProgramDefinition().adminName();
+                    },
+                    Collectors.groupingBy(
+                        Application::getLifecycleStage,
+                        // In practice, we don't expect an applicant to have multiple
+                        // DRAFT or ACTIVE applications for a given program. Grabbing the latest
+                        // application here guards against that case, should it occur.
+                        Collectors.maxBy(Comparator.comparingLong(a -> a.id)))));
+
+    ImmutableList.Builder<ApplicantProgramData> inProgressPrograms = ImmutableList.builder();
+    ImmutableList.Builder<ApplicantProgramData> submittedPrograms = ImmutableList.builder();
+    ImmutableList.Builder<ApplicantProgramData> unappliedPrograms = ImmutableList.builder();
+
+    Set<String> programNamesWithApplications = Sets.newHashSet();
+    mostRecentApplicationsByProgram.forEach(
+        (programName, appByStage) -> {
+          Optional<Application> maybeDraftApp =
+              appByStage.getOrDefault(LifecycleStage.DRAFT, Optional.empty());
+          Optional<Application> maybeSubmittedApp =
+              appByStage.getOrDefault(LifecycleStage.ACTIVE, Optional.empty());
+          Optional<Instant> maybeSubmitTime = maybeSubmittedApp.map(Application::getSubmitTime);
+          if (maybeDraftApp.isPresent()) {
+            inProgressPrograms.add(
+                ApplicantProgramData.create(
+                    maybeDraftApp.get().getProgram().getProgramDefinition(), maybeSubmitTime));
+            programNamesWithApplications.add(programName);
+          } else if (maybeSubmittedApp.isPresent() && activeProgramNames.containsKey(programName)) {
+            ProgramDefinition programDef = activeProgramNames.get(programName);
+            submittedPrograms.add(ApplicantProgramData.create(programDef, maybeSubmitTime));
+            programNamesWithApplications.add(programName);
+          }
+        });
+
+    Set<String> missingActivePrograms =
+        Sets.difference(activeProgramNames.keySet(), programNamesWithApplications);
+    missingActivePrograms.forEach(
+        programName -> {
+          unappliedPrograms.add(
+              ApplicantProgramData.create(activeProgramNames.get(programName), Optional.empty()));
+        });
+
+    // Ensure each list is ordered by database ID for consistent ordering.
```

Comment:
Per the above comment, the program.id is not the most correct ordering given the system version can be reverted.

Reply:
Agreed. That said, I think this is the current behavior (which I was attempting to preserve).

We gather the list of active programs based on "natural order" from the active version (e.g. no ordering applied to the query) ([code](https://github.com/seattle-uat/civiform/blob/e1f02bb0b6975f308745365a40ad32203e830ef0/server/app/models/Version.java#L78-L80)). The natural order for the many to many would be the Primary Key, which uses the program ID ([code](https://github.com/seattle-uat/civiform/blob/e1f02bb0b6975f308745365a40ad32203e830ef0/server/conf/evolutions/default/17.sql#L19)).

Given this, would you prefer we use a different ordering? I'm not sure what we'd want to use since there's not a clear signal. Using alphabetical seems like it'd be a big change from what we have now.